### PR TITLE
Add support for persistent superchats

### PR DESCRIPTION
**Pull request by KentoNishi**
Implements https://github.com/LiveTL/HyperChat/issues/49

image
image

### CODE REVIEW

**Review by r2dev2 — CHANGES_REQUESTED** [review]
The superchat bar is in light theme when yt is in dark theme when "use yt theme" is selected. I just did `yarn start:firefox` for this. The theme of the superchat bar works fine for manually set themes however. Everything else seems to be fine.

superchat bar in light theme with yt in dark theme

**Review by r2dev2 — APPROVED** [review]

File: src/components/Settings.svelte [hunk 1]
```diff
@@ -16,17 +16,11 @@
...
     background: transparent;
   }
-  :global(::-webkit-scrollbar-thumb) {
-    background: #888;
-  }
-  :global(::-webkit-scrollbar-thumb:hover) {
-    background: #555;
+  * {
+    scrollbar-width: thin;
+    scrollbar-color: #888 transparent;
```
Comment (ChrRubin) [inline src/components/Settings.svelte:24]: `webkit-scrollbar-track` probably isn't needed if thin scrollbars are removed. Also if we're removing thin scrollbars here on Chrome, we should also remove it on Firefox.

File: src/components/SuperchatViewDialog.svelte [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  $: sc = $focusedSuperchat as Ytc.ParsedTimedItem;
+</script>
+
+<Dialog active={Boolean(sc)} noCloseButton>
+  <PaidMessage message={sc}  />
+  <div slot="actions">
+    <Button on:click={() => {
+      $focusedSuperchat = null;
+    }} color="primary">Close</Button>
+  </div>
```
Comment (ChrRubin) [inline src/components/SuperchatViewDialog.svelte:17]: The close button feels unnecessary.

**Review by ChrRubin — CHANGES_REQUESTED** [review]
Some extra stuff aside from the comments:

- The ticker bar seems to just disappear randomly and never come back? I experienced this during Watame's BDay live on both Firefox and Chromium. Seems to only affect live streams and not VODs. No errors in the console.

- If we want to use name for the ticker items instead of profile pic like YTC does, we might want to truncate it in case of people with long names taking up too much space like this one:

image

- Membership items (not talking about gifts) and membership milestone chats aren't being stickied even though they are on normal YTC. 

- Adding a setting to toggle the ticker bar would be nice.

**Comment by ChrRubin** [comment]
Ran with Astel's BDay live and now the ticker bar is always there but new superchats aren't added to it lmao

**Comment by KentoNishi** [comment]
> Membership items (not talking about gifts) and membership milestone chats aren't being stickied even though they are on normal YTC.

this is intentional, imo they dont belong in the superchat bar and it just gets in the way of actual messages that need to be stickied so imma leave that as it is

**Comment by KentoNishi** [comment]
altho ig there are membership joins with messages, we could add that later possibly

**Comment by KentoNishi** [comment]
@ChrRubin it took a while, but i implemented all the suggestions and improved stuff all around. pls review thx

**Comment by KentoNishi** [comment]
image

**Comment by KentoNishi** [comment]
hmmmmmm odd, the issue of new ticker items not being added in livestreams still happens...

**Comment by KentoNishi** [comment]
ok fixed it, the issue was that the superchat was added but was getting deleted immediately because the poll time wasn't caught up. pls try now @ChrRubin

**Review by ChrRubin — CHANGES_REQUESTED** [review]
Aside from the absolute div, seems good

File: src/components/Hyperchat.svelte [hunk 3]
```diff
@@ -289,18 +322,22 @@
...
+  <div class="absolute top-0 w-full">
+    {#if $enableStickySuperchatBar}
+      <StickyBar />
+    {/if}
+    {#if pinned}
+      <div class="mx-2 mt-2">
+        <PinnedMessage pinned={pinned} />
+      </div>
+    {/if}
+  </div>
```
[the remark this answers is not in the document]
Comment (KentoNishi) [reply src/components/Hyperchat.svelte:340]: this is the same with pinned messages btw.

**Comment by KentoNishi** [comment]
@ChrRubin 

https://user-images.githubusercontent.com/38841491/173536345-52f054e6-3356-45ea-9334-8fcc6fb14383.mp4

**Comment by ChrRubin** [comment]
LMAO I assumed you were gonna just put the sticky bar as a separate div above the "main" HC div like how YTC does it.

Could you try to do that instead, or is there some other problem preventing that? Seems to me like it'll involve a lot less fucking around with the heights and stuff and will fix this OCD trigger:

https://user-images.githubusercontent.com/20059164/174049354-b1124563-5095-44b5-ab5a-4b8b371b279b.mp4

File: src/scripts/chat-background.ts [hunk 4]
```diff
@@ -428,9 +428,7 @@ chrome.runtime.onMessage.addListener((request, sender, sendResponse) => {
   } else if (request.type === 'createPopup') {
     chrome.windows.create({
       url: request.url,
-      type: 'popup',
-      height: 420,
-      width: 690
+      type: 'popup'
     }, () => {});
```
Comment (ChrRubin) [inline src/scripts/chat-background.ts:432]: We sure we wanna remove the width/height? The defaults seems a tad large.

image

**Comment by ChrRubin** [comment]
I've fixed the OCD trigger in ba7bc8d, so the sticky bar is no longer in an absolute div. Pinned messages not blocking the top few messages is nice so we should still keep that part regardless.

For the popup, I think we should still specify a certain size instead of letting the browser wing it and fill up half of the screen. Maybe slightly larger than before if needed.

**Comment by KentoNishi** [comment]
@ChrRubin chrome now inconsistently enforces a minimum pop-up window area size based on the percentage of area it occupies. I realized this while deving on my laptop -- chrome throws an error if a window takes less than 50% of available screen area, blocking the pop-up. Getting rid of the dimensions is the easiest solution, unless u wanna try messing with measuring screen dimensions and sht lol

**Comment by KentoNishi** [comment]
btw @ChrRubin here's the error, screenshotted from the production chrome listing

image

**Comment by ChrRubin** [comment]
> btw @ChrRubin here's the error, screenshotted from the production chrome listing
> 
> image
> 

Jesus that's stupid. No wonder the default literally just fills up half of the screen on Chrome.

**Review by ChrRubin — APPROVED** [review]
@KentoNishi you should double-check that my commit doesn't mess up any of the stuff you changed. Seems good to me at quick glance

**Comment by KentoNishi** [comment]
Got it, will do after I get back to mah hotel

**Pull request merged**